### ISSUE TITLE
minimal mak:list and mak:form APIs

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#771]
**Reported by @cristianbogdan on 16 May 2010 22:01 UTC**
the job of mak:list, given a data source, is to
- compose queries (simple concatenation in principle)
- figure out the projection types.
- get the data, group it, and show the mak:values based on the types.

it strikes me that you don't necessarily need MDDs to do this. there can be a much simpler interface providing all that's needed for makumba to analyze a page and work over query results. 

like 
- if you have a hibernate query you don't need MDDs to figure out the return projections. you only need hibernate mappings. 
- if you have a mysql database, you only need SHOW CREATE TABLE
- i guess the same goes for JPA and any other data source.

even if makumba actually manage the query language, analyzing these queries needs about the same information as above. like you could do MQL over a mysql database without having any MDDs about it, just its tables, with foreign keys.

in order to use the current query analysis and mak:list, we would need to transform all this info into the MDD interfaces. while that can be easy to do and opens the door to a lot of functionality that we already have and is based on MDDs,  it also brings limitations such as not having inheritance, having a limited type set, etc.

my hunch here is that the interface needed to implement the above, would be a much smaller one than the MDD interface
- return a [Sorted]Map<String, Class>  (or simply Class if that's easier!) for query projections and another for query parameters
- for that, given a type (Class) t, tell whether t.field is a reference to another class or a set of another class (at query analysis level, it doesn't matter whether this is a subpointer or subset). 

so it may be easier in the end to make the query analyzers and the page analyzers to work with such very-small interfaces, and then we could implement such an interface based on the current MDD world.

going further, we would actually not even need our CRUD API, to implement mak:form, but define a much smaller interface (i think we already have something of the kind), which the respective layer could implement, and maybe a hook like the onXXX methods  to give an event-driven possibility for the programmer to add to the actions of the makumba framework. and again, the current code can become an implementation of that.

so basically this proposal is to decouple mak:list and mak:form from the MDD API and Dictionary CRUD APIs.

of course, in the MDD side there are the rules and the functions. i think they are a primary strength of makumba and we should of course have them.  but this proposal makes them independent on how the fields are defined. whether they are written in separate files or as annotations of the original files, is less important.

but for the rest, makumba can be made accessible for those who don't want to use:
- .mdd files
- the dictionary API
- maybe not even MQL

also, a larger number of technologies can be made accessible to the makumba world through these smaller interfaces.

Migrated-From: http://trac.makumba.org/ticket/1226

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 17 May 2010 07:02 UTC**
(In reply to comment #0)
> the job of mak:list, given a data source, is to
> - compose queries (simple concatenation in principle)
> - figure out the projection types.
> - get the data, group it, and show the mak:values based on the types.
> 
> it strikes me that you don't necessarily need MDDs to do this. there can be a
> much simpler interface providing all that's needed for makumba to analyze a
> page and work over query results.

I agree that it can be simpler for that purpose. MDDs are most heavily used in the database layer, where the full meta-model of relationships needs to be known (to generate the schema). see bug 927 also.

> my hunch here is that the interface needed to implement the above, would be a
> much smaller one than the MDD interface
> - return a [Sorted]Map<String, Class>  (or simply Class if that's easier!) for
> query projections and another for query parameters
> - for that, given a type (Class) t, tell whether t.field is a reference to
> another class or a set of another class (at query analysis level, it doesn't
> matter whether this is a subpointer or subset). 


this proposal looks similar to 
http://java.sun.com/javaee/6/docs/api/javax/persistence/metamodel/ManagedType.html

I guess it's worth looking at that design.


> so it may be easier in the end to make the query analyzers and the page
> analyzers to work with such very-small interfaces, and then we could implement
> such an interface based on the current MDD world.

when thinking of a JSF mak:list I am actually not considering a page analysis anymore but an API that lets you do the list composition / grouping and that would be used by the JSP taglib (with page analysis) and the JSF API (with knowledge of the whole page tree at execution). that way the same engine can be used for many clients.

> so basically this proposal is to decouple mak:list and mak:form from the MDD
> API and Dictionary CRUD APIs.

the most challenging part of the decoupling as I see it now is that we have a lot of MDD provider calls all over the place. I mean we don't have much control over where it is being used, as many classes just do DataDefinitionProvider.getInstance().doSomething(). this is what the spring / google clean code folks critizise (their approach is that this kind of data - the MDD model or MetaModel or whatever - should be passed via constructors or calling methods, not taken statically from a singleton. it also makes testing hard, if not impossible.
 
> of course, in the MDD side there are the rules and the functions. i think they
> are a primary strength of makumba and we should of course have them.  but this
> proposal makes them independent on how the fields are defined. whether they are
> written in separate files or as annotations of the original files, is less
> important.

my proposal for Java POJOs:

@QueryFragment('this.name + ' ' + this.surname')
public void fullName() {}

or  la GWT:

public native void fullName() { /* this.name + ' ' + this.surname  */ }

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 May 2010 11:11 UTC**
(In reply to comment #1)
> I agree that it can be simpler for that purpose. MDDs are most heavily used in
> the database layer, where the full meta-model of relationships needs to be
> known (to generate the schema). see bug 927 also.

you don't need them there either! you can generate a Class or SortedMap<String, Class> with fields in the right order, and with annotations for notnull and unique (or simply adding validation rules dynamically for those). then you isolate the table thing from DataDefinition completely. you only have SortedSet/Class and the rules.

in the end DataDefinition is just a parser for the MDD format and a  translator to SortedMap/Class + rules. one consequence is that you won't be able to reload an MDD without reloading the webapp/classloader but i don't think that's a big problem. 

important consequence: no more MDD parser for anything else than .mdd. parsers maybe (for things like mysql SHOW CREATE TABLE), but produce directly Class/SortedMap/ManagedType. so no parser for JPA and Hibernate as they already rely on Class :)

one special issue will be intEnum and charEnum... i guess we'll have to generate some java enumerators for them

a pointer simply becomes a field of the respective type, or some sort of Reference to it
a set becomes a Collection of the respective type, or a collection of References.

> this proposal looks similar to 
> http://java.sun.com/javaee/6/docs/api/javax/persistence/metamodel/ManagedType.html
> 
> I guess it's worth looking at that design.

great point(er). maybe we could stick to Class or SortedMap wherever possible because we can always produce a SortedMap from a class, or a Class (or SortedMap) from ManagedType but maybe not the other way around. not sure. 

certainly for query result types i wouldn't want to generate a new class in the classloader, i'd to a SortedSet<String, Class> (or even unsorted actually!)

> when thinking of a JSF mak:list I am actually not considering a page analysis
> anymore but an API that lets you do the list composition / grouping and that

well the syntactics of the page analysis for JSF and JSP are different because JSF allows you to get the children and JSP not. but the semantics remain makumba-specific and we should change as little of that as possible.

> the most challenging part of the decoupling as I see it now is that we have a
> lot of MDD provider calls all over the place. I mean we don't have much control

well if you look at mak:list, it uses DataDefinition in very few places really. same goes for query providers (especially if we separate FunctionDefinition from DataDefinition per se). maybe mak:form needs a bit more and certainly it needs CRUD. 

> over where it is being used, as many classes just do
> DataDefinitionProvider.getInstance().doSomething(). this is what the spring /

oh those will have to disappear. and we can do it gradually, like 
- separate FunctionDefinition and ValidationDefinition from DataDefinition (DD) and FieldDefinition (FD)
- make (revive) the MDD2Class generator and make a provider. given a type name like best.Student return best.Student.class. if we can examine the MDDs when the application loads, we can simply use Class.forName to retrieve the types! this is beautiful. 
- the big issue above is whether we are able to produce class files with <template params>. if not, we can fix SortedMap<String, Class> and come with some solution for pointers.
- remove DD and FD from query providers. retrieve a Class (instead of DD) with the provider above when you need a type, and look at its fields
- remove DD and FD from mak:list (really easy once removed from query providers)
- remove DD and FD from the CRUD Dictionary API and implementation (table)
- remove DD and FD from mak:form

so when do we start? :)

> my proposal for Java POJOs:
> 
> @QueryFragment('this.name + ' ' + this.surname')
> public void fullName() {}

> 
> or  la GWT:
> 
> public native void fullName() { /* this.name + ' ' + this.surname  */ }

great pointers again. i guess we can start with an own RFD format (rules and function definition :) ) with basically the same content as MDDs today but without fields, while keeping in mind the other possibilities.

or we can decide that if you do MDDs (and thus have no classes as such) you put functions in the MDD file, if not you have to have them as annotations or so. i guess annotations you can validate in IDEs so there may be "direct" programming support there

**Comment by ghost** [comment]
**Modified by @cristianbogdan on 17 May 2010 11:13 UTC**

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 May 2010 11:25 UTC**

so we have to look at these in the MDD class generator:

> one special issue will be intEnum and charEnum... i guess we'll have to
> generate some java enumerators for them
http://java.sun.com/j2se/1.5.0/docs/guide/language/enums.html

so 

en= int{"a"=1, "b"=2}

could generate Java like

public enum en_enum extends MakumbaIntEnum {
  a ("a", 1),
  b ("b", 2)
};

en_enum en;

> a pointer simply becomes a field of the respective type, or some sort of
> Reference to it
> a set becomes a Collection of the respective type, or a collection of
> References.

....

> - the big issue above is whether we are able to produce class files with
> <template params>. if not, we can fix SortedMap<String, Class> and come with
> some solution for pointers.

i think pointers are actually no problem. the problem is sets

p= ptr T generates the java
T p;
s= set T generates the java
Collection<T> s;

and if solving the generic type T above, we can also generate a reference for pointers
Reference<T> p;

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 May 2010 15:22 UTC**
so i looked at MddToJava and saw it's made with javassist, a part of jboss. 
the good news is, javassist is well maintained!

http://www.csg.is.titech.ac.jp/~chiba/javassist/

(In reply to comment #4)
> en= int{"a"=1, "b"=2}
> 
> could generate Java like
> 
> public enum en_enum extends MakumbaIntEnum {
>   a ("a", 1),
>   b ("b", 2)
> };
> 
> en_enum en;


enums seem to be covered by javassist

http://kickjava.com/src/javassist/bytecode/annotation/EnumMemberValue.java.htm

there seem to be some problems with them but essentially they are supported.

we already did an attempt to generate eums with javassist but the generateIntEnum() method is not called anywhere so i guess it fails.

javassist seems to have different API levels and it may be that CtClass, the level we use, does not support everything that the lower levels support.

> i think pointers are actually no problem. the problem is sets
> 
> p= ptr T generates the java
> T p;
> s= set T generates the java
> Collection<T> s;
> 
> and if solving the generic type T above, we can also generate a reference for
> pointers
> Reference<T> p;

javassist is supposed to support generics
http://www.csg.is.titech.ac.jp/~chiba/javassist/tutorial/tutorial3.html#generics

but according to the page, it seems that the java bytecode per se does not contain generic information (and that makes sense actually, it is only used at compile time to check types). however, java.lang.Class does contain generic parameter info, so maybe they are included in the type info (not in the method bytecode though)

http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Class.html#getTypeParameters()

in the worst case, javassist supports annotations, maybe we can use that (generate a Collection subclass and annotate it with the type)

the Makumba set generates a subtable class and i think that's not needed for this purpose. but we can have it anyway. for subsets it's clearly needed but we have to remember to add also a Collection<subsetClass> like for normal sets we add Collection<T>

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 17 May 2010 17:12 UTC**
(In reply to comment #5)
> we already did an attempt to generate eums with javassist but the
> generateIntEnum() method is not called anywhere so i guess it fails.

(In reply to comment #2)
> > over where it is being used, as many classes just do
> > DataDefinitionProvider.getInstance().doSomething(). this is what the spring /
> 
> oh those will have to disappear. and we can do it gradually, like 
> - separate FunctionDefinition and ValidationDefinition from DataDefinition (DD)
> and FieldDefinition (FD)
> - make (revive) the MDD2Class generator and make a provider. given a type name
> like best.Student return best.Student.class. if we can examine the MDDs when
> the application loads, we can simply use Class.forName to retrieve the types!
> this is beautiful. 
> - the big issue above is whether we are able to produce class files with
> <template params>. if not, we can fix SortedMap<String, Class> and come with
> some solution for pointers.
> - remove DD and FD from query providers. retrieve a Class (instead of DD) with
> the provider above when you need a type, and look at its fields
> - remove DD and FD from mak:list (really easy once removed from query
> providers)
> - remove DD and FD from the CRUD Dictionary API and implementation (table)
> - remove DD and FD from mak:form

about validation rules: let us not forget about JSR 303 :-)

http://relation.to/Bloggers/BeanValidationSneakPeekPartI

my "plan" (or thoughts) for JSF is to generate those from the MDD validation rules. then of course you still have the makumba VDs for JSP. in any case I would look into having a mapping between MDD VRs and these bean-based validation rules.

(note that in the demo app you can see these rules at work, if I am not mistaken. they directly print red messages in the JSF <h:message> tags)

> so when do we start? :)

ASAP I would say :) I mean this is exactly what is needed for introducing JSF, to have a bridge to the POJO/EJB world.

one part where design is not clear to me though is the makumba BL. we have all these on_topOfSomething() which work well in the JSP-based one-request world, but with AJAX and its dozens of requests that may not work that well. I guess this is food for another entry though.

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 17 May 2010 17:14 UTC**
now is one of the times I would like to hit "edit" on the comment:

(In reply to comment #6)
> (In reply to comment #5)
> > we already did an attempt to generate eums with javassist but the
> > generateIntEnum() method is not called anywhere so i guess it fails.

so I was going to say: javassist did not generate enums at the time so this method was failing. maybe it works now, it should be tested.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 May 2010 17:40 UTC**
i played a bit with javassist and there's no hope for generics. they are simply
not existing at bytecode level.

so i think we should use standard JPA annotations for them

http://java.sun.com/javaee/6/docs/api/javax/persistence/ManyToMany.html

more precisely

  @ManyToMany(targetEntity=com.acme.PhoneNumber.class)
    public Set getPhones() { return phones; }

javassist should generate annotations so that's no problem.


----

the enum business, i don't know yet.  please help.  we have to crack this one
:)

there is the Enumerated annotation in javax.persistence, but it can only save
STRING or ORDINAL. we want to save something else. sure we are not interested
in what's saved as yet because we do the saving, we only need the type. maybe
later we can make some JPA transformer (or whatever, no clue) to save the right
value

also javassist does not support inner classes, so we would have to generate a
separate class

public enum best.Student_someIntEnum{
       a(1), b(2), c(3);

        private final int value;

        Student_someIntEnum(int v) {
            this.value = v;
        }
}

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 17 May 2010 17:57 UTC**
(In reply to comment #8)
> i played a bit with javassist and there's no hope for generics. they are simply
> not existing at bytecode level.
> 
> so i think we should use standard JPA annotations for them
> 
> http://java.sun.com/javaee/6/docs/api/javax/persistence/ManyToMany.html
> 
> more precisely
> 
>   @ManyToMany(targetEntity=com.acme.PhoneNumber.class)
>     public Set getPhones() { return phones; }
> 
> javassist should generate annotations so that's no problem.

this looks like a good solution, I would not know what else could be used. maybe there is a way of populating the Class method you were referring to earlier?


> ----
> 
> the enum business, i don't know yet.  please help.  we have to crack this one
> :)

http://stackoverflow.com/questions/2751733/map-enum-in-jpa-with-fixed-values

seems to present a number of solutions, I guess we can pick one of them

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 May 2010 19:56 UTC**
found this list of bytecode editors. heard about cglib 

http://www.java2s.com/Product/Java/Byte-Source-Code/Byte-Code-Edit.htm

(In reply to comment #9)
> this looks like a good solution, I would not know what else could be used.
> maybe there is a way of populating the Class method you were referring to
> earlier?

no i think JPA is good enough. 
i think we can generate JPA annotations for anything else we can, like for "not null", etc

> http://stackoverflow.com/questions/2751733/map-enum-in-jpa-with-fixed-values
> 
> seems to present a number of solutions, I guess we can pick one of them


well javassist does not generate enums at all, whether as inner classes or not.
http://www.csg.is.titech.ac.jp/~chiba/javassist/tutorial/tutorial2.html#limit

also it generates annotations in a very shitty way (low-level API)
http://www.csg.is.titech.ac.jp/~chiba/javassist/html/javassist/bytecode/annotation/AnnotationsWriter.html

it also feels to me like we use very few javassist features which might mean it's not the most suitable here. so we have to examine the list above for:
- on-the-fly (no javac) class generation, not necessarily change existing classes i think
- easy generation of class and field annotations
- maybe type parameters to methods, if there is any hope (i don't know what javassist means by "the lower-level API of Javassist fully supports generics introduced by Java 5." i certainly was not able to retrieve the generic (String) from
ArrayList<String> x= new ArrayList<String>();
x.getClass().get....
i could find no method that tells me more than "x is of type ArrayList". nothing about String. but maybe there is....

and even if we use that, we have to look for java annotations as well.

------

about how to do with enums. in general i think we have a number of constraints here:
- reduce the use of DD and FD to a minimum (i.e. to a conversion to Class and work with Class all over makumba)
- make sure that present DBs will work with present MDD code
- clearly JPA has no good solution for intEnum, and neither seems to have Hibernate. i remember seeing genericEnumUserType before and i think we attempted to get rid of it when we tried to generate an enum

that is, we do not have at this point to make sure that _any_ JPA application will generate precisely the same db and interface content as the makumba intEnum. for now we only want to define the data type somehow, not its instances (?) or database layer

so my suggestion is a custom annotation like

@MakEnum("a"=1, "b"=2)
int fieldName;

this allows us to go on removing references to MDDs

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 May 2010 20:22 UTC**
about instances of the MDD classes (replacements for dictionaries, simple POJO). for simple fields, it's easy. the "only" problem comes for pointers.

i guess we can go on like this:

best.Student student;

public best.Student getStudent(){ ... )

initially the field is just an empty best.Student, with just the id populated. so the  traditional Pointer value can be obtained like

getStudent().getId()

the moment we request another field

getStudent().getName()

then a read is requested from the db to populate the other fields.

this is then a minimal POJO.

the same happens if we return a Map<String, Object> from a query.but maybe there we can use directly the JPA Query API

http://java.sun.com/javaee/6/docs/api/javax/persistence/Query.html

or at least Tuple for the result row
http://java.sun.com/javaee/6/docs/api/javax/persistence/Tuple.html

in any case the Student from such a tuple would be just a lazy one as illustrated above.

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 17 May 2010 21:24 UTC**
(In reply to comment #11)
> public best.Student getStudent(){ ... )
> 
> initially the field is just an empty best.Student, with just the id populated.
> so the  traditional Pointer value can be obtained like
> 
> getStudent().getId()
> 
> the moment we request another field
> 
> getStudent().getName()
> 
> then a read is requested from the db to populate the other fields.

I am not entirely sure if I understand the problem here.

All you need to ensure is that if you want laziness, the one-to-one relationship between best.Lbg and best.Student is lazy. but I think there are ways to configure this globally for all one-to-one's.

for the rest, all those beans are then just managed by JPA which takes care of any kind of optimizations necessary.

the crux is to have these generated entity beans registered as managed beans to JPA, but that is not too much of an issue, we have it already with hibernate, I think.


the javassist stuff is more annoying. it might be an idea to look into ASM (http://asm.ow2.org/) or cglib (http://cglib.sourceforge.net/) if they can do better. otherwise, there might be means to compile and load a class into the JVM at runtime...I mean I am sure that in J2EE environments they do this, I will ask a colleague tomorrow. that would mean the first startup is slower but once the files are compiled it should be fine.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 May 2010 22:04 UTC**
(In reply to comment #12)
> I am not entirely sure if I understand the problem here.
> 
> All you need to ensure is that if you want laziness, the one-to-one
> relationship between best.Lbg and best.Student is lazy. but I think there are
> ways to configure this globally for all one-to-one's.

> for the rest, all those beans are then just managed by JPA which takes care of
> any kind of optimizations necessary.

the problem/constraint would be not to commit to JPA entity management, at least not in the first instance. like do the minimal work needed to get makumba to work without calling into MDD code all over the place.

i have a hunch that BEST will still want to use the Dictionary API and the makumba CRUD implementation and schema update. no JPA. like it happened with hibernate.

but if the query functions reduce the BL to a level where it can be easily ported to some light JPA implementation, we can talk. but i have my reserves there, at least for the beginning.

that's why i proposed a very very simple POJO, from which we can make a Dictionary when needed (and the other way around). the POJO interface will already look like it would look in the JPA/Hibernate world. so internally makumba would look at POJOs (or at least Map<String, Object> or Tuple) as it will do in the future (JPA/Hibernate) but it will be able to 
1) obtain these POJOs or tuples without JPA or Hibernate if needed
2) return Dictionaries if needed

that's the problem/rationale :)

> the crux is to have these generated entity beans registered as managed beans to
> JPA, but that is not too much of an issue, we have it already with hibernate, I
> think.

yes but not in the legacy karamba case.

> the javassist stuff is more annoying. it might be an idea to look into ASM
> (http://asm.ow2.org/) or cglib (http://cglib.sourceforge.net/) if they can do
> better. otherwise, there might be means to compile and load a class into the
> JVM at runtime...I mean I am sure that in J2EE environments they do this, I
> will ask a colleague tomorrow. that would mean the first startup is slower but
> once the files are compiled it should be fine.

sure it's no problem to compile code. but the point was to avoid that... the compiler is a heavy beast, and you end up with N more files (java files). on the other hand in a JSP environment there is a compiler anyway, it's just a matter of finding it.

and remember that the slow hibernate startup is partly what doomed the whole thing in the hibernate case. it was mostly  coming from the schema update, but still.

whatever class generation we use, we should make sure it is incremental (we do it when we need the type, not at once for all types). also if we can find a JPA implementation with incremental schema update, that would be great.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 May 2010 23:53 UTC**
javassist worked in the end, the main() of MddToClass does a JPA annotation

i thus think we have solved set and ptr

i think we for now we can use a makumba-specific annotation for intEnum in order to make progress. then we can recognize other annotations/constructs at the view level (and query parameter value level?)  to support enumerators in other ways. since we have total control of the code we generate, if we find a better construct later,  we can generate that one and honor it at view level (and query param?)

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 18 May 2010 10:12 UTC**
(In reply to comment #13)
> that's why i proposed a very very simple POJO, from which we can make a
> Dictionary when needed (and the other way around). the POJO interface will
> already look like it would look in the JPA/Hibernate world. so internally
> makumba would look at POJOs (or at least Map<String, Object> or Tuple) as it
> will do in the future (JPA/Hibernate) but it will be able to 
> 1) obtain these POJOs or tuples without JPA or Hibernate if needed
> 2) return Dictionaries if needed
> 
> that's the problem/rationale :)

I'm a bit skeptical of going down that road because that sounds awfully a lot like implementing a ORM solution. I mean for read-only it is fine. but as soon (and this will probably happen very soon) as someone wants to modify an object state you have to track this somehow. this is what all the idiomatic java is about. they use deep and sophisticated introspection methods to figure out what changed in the state of an object.

one example where I see this could happen

on_editBestStudent(best.Student student, Transaction db, Attributes a) {
  new Bungee(student).jump();
}

the above Bungee jump() changes the state of the student object, and that could be in many sophisticated ways such as student.getHormonalSystem().setAdrenalineLevel(1000);

with the current Dictionary API such things are possible (and occur in K!). So I think the ability to set stuff will be demanded by the users or it renders the whole thing rather impossible to use.

And perhaps there are ways of implementing it - but I am afraid that this is hard. I mean the hibernate folks spent years to make their idiomatic engine good enough, I don't think this should be our main focus.


I know the problem of karamba and speed of loading. Before going into the design of a makumba-ORM-layer though I would:
- investigate how JPA manages schema update, if they have the incremental one
- make a speed test with K! (also simply by updating the hibernate version, maybe they improved on that)
- see http://www.informit.com/articles/article.aspx?p=353737 and search for "incremental", there seems to be a tool to upgrade just a set of types in the schema, that can be talked to programmatically.

What about looking into using JPA as a database layer and implement a dictionary -> JPA bridge. that way K! still uses dictionary in many places and can use the JPA POJOs in new places.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 18 May 2010 10:40 UTC**
i think we can do this gradually
- MDD -> POJO class (bug 1229, also bug 1227 and bug 1228 but with low prio)
- query API -> JPA query API (bug 1230)
- hierarchical Dictionary -> POJO graph (bug 1231)

all of the above allow us to "plug in" to JPA later on while still not needing
to commit to JPA yet. we can have working code at all stages.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 19 May 2010 15:15 UTC**
this contains a lot of intermediary discussion and speculation whose results are distilled in other entries, as per comment 16
